File: src/main/java/me/hhitt/disasters/arena/service/ResetArenaService.java
package me.hhitt.disasters.arena.service;

import com.sk89q.worldedit.WorldEditException;
import com.sk89q.worldedit.bukkit.BukkitAdapter;
import com.sk89q.worldedit.bukkit.WorldEditPlugin;
import com.sk89q.worldedit.extent.clipboard.BlockArrayClipboard;
import com.sk89q.worldedit.extent.clipboard.Clipboard;
import com.sk89q.worldedit.function.operation.ForwardExtentCopy;
import com.sk89q.worldedit.function.operation.Operations;
import com.sk89q.worldedit.math.BlockVector3;
import com.sk89q.worldedit.regions.CuboidRegion;
import com.sk89q.worldedit.session.ClipboardHolder;
import me.hhitt.disasters.Disasters;
import me.hhitt.disasters.arena.Arena;
import me.hhitt.disasters.game.GameState;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.entity.Player;

import java.util.logging.Logger;

public class ResetArenaService {

    private static final long WARN_THRESHOLD_MS = 5000L;

    private final Arena arena;
    private final WorldEditPlugin worldEdit;
    private final World world;
    private final Logger logger;
    private Clipboard clipboard;
    private BlockVector3 center;
    private boolean clipboardInitialized;

    private final int minX;
    private final int maxX;
    private final int minY;
    private final int maxY;
    private final int minZ;
    private final int maxZ;

    public ResetArenaService(final Arena arena, final WorldEditPlugin worldEdit) {
        this.arena = arena;
        this.worldEdit = worldEdit;
        this.world = arena.getCorner1().getWorld();
        this.logger = Disasters.getInstance().getLogger();

        this.minX = (int) Math.min(arena.getCorner1().getX(), arena.getCorner2().getX());
        this.maxX = (int) Math.max(arena.getCorner1().getX(), arena.getCorner2().getX());
        this.minY = (int) Math.min(arena.getCorner1().getY(), arena.getCorner2().getY());
        this.maxY = (int) Math.max(arena.getCorner1().getY(), arena.getCorner2().getY());
        this.minZ = (int) Math.min(arena.getCorner1().getZ(), arena.getCorner2().getZ());
        this.maxZ = (int) Math.max(arena.getCorner1().getZ(), arena.getCorner2().getZ());
    }

    public void save() {
        final World world = arena.getCorner1().getWorld();
        if (world == null) {
            logger.severe("World is null before save!");
            return;
        }
        final BlockVector3 min = BlockVector3.at(arena.getCorner1().getX(), arena.getCorner1().getY(), arena.getCorner1().getZ());
        final BlockVector3 max = BlockVector3.at(arena.getCorner2().getX(), arena.getCorner2().getY(), arena.getCorner2().getZ());
        final CuboidRegion region = new CuboidRegion(min, max);
        final long start = System.nanoTime();
        logger.info("Saving arena '" + arena.getName() + "' blocks: " + region.getVolume());

        final BlockArrayClipboard clip = new BlockArrayClipboard(region);
        try (com.sk89q.worldedit.EditSession editSession = worldEdit.getWorldEdit().newEditSession(BukkitAdapter.adapt(world))) {
            final ForwardExtentCopy forwardExtentCopy = new ForwardExtentCopy(editSession, region, clip, region.getMinimumPoint());
            Operations.complete(forwardExtentCopy);
            this.clipboard = clip;
            this.center = region.getMinimumPoint();
            this.clipboardInitialized = true;
        } catch (WorldEditException e) {
            throw new RuntimeException(e);
        }
        final long elapsed = (System.nanoTime() - start) / 1_000_000L;
        if (elapsed >= WARN_THRESHOLD_MS) {
            logger.warning("Save completed for '" + arena.getName() + "' in " + elapsed + " ms (slow)");
        } else {
            logger.info("Save completed for '" + arena.getName() + "' in " + elapsed + " ms");
        }
    }

    public void paste() {
        if (!clipboardInitialized) {
            logger.severe("Clipboard not initialized!");
            return;
        }
        logger.info("Paste triggered for '" + arena.getName() + "'");
        final long start = System.nanoTime();
        removeEntitiesInRegion();

        try (com.sk89q.worldedit.EditSession editSession = worldEdit.getWorldEdit().newEditSession(BukkitAdapter.adapt(world))) {
            logger.info("Pasting arena '" + arena.getName() + "' now...");
            final com.sk89q.worldedit.function.operation.Operation operation = new ClipboardHolder(clipboard)
                .createPaste(editSession)
                .to(center)
                .ignoreAirBlocks(false)
                .build();
            Operations.complete(operation);
        } catch (WorldEditException e) {
            throw new RuntimeException(e);
        }
        refreshChunks(world, arena.getCorner1(), arena.getCorner2());
        arena.setState(GameState.RECRUITING);
        final long elapsed = (System.nanoTime() - start) / 1_000_000L;
        if (elapsed >= WARN_THRESHOLD_MS) {
            logger.warning("Paste completed for '" + arena.getName() + "' in " + elapsed + " ms (slow)");
        } else {
            logger.info("Paste completed for '" + arena.getName() + "' in " + elapsed + " ms");
        }
    }

    private void refreshChunks(final World world, final Location loc1, final Location loc2) {
        final int minXX = minX >> 4;
        final int maxXX = maxX >> 4;
        final int minZZ = minZ >> 4;
        final int maxZZ = maxZ >> 4;

        for (int x = minXX; x <= maxXX; x++) {
            for (int z = minZZ; z <= maxZZ; z++) {
                world.refreshChunk(x, z);
            }
        }
    }

    private void removeEntitiesInRegion() {
        for (org.bukkit.entity.Entity entity : world.getEntities()) {
            final Location loc = entity.getLocation();
            if (loc.getBlockX() >= minX && loc.getBlockX() <= maxX
                && loc.getBlockY() >= minY && loc.getBlockY() <= maxY
                && loc.getBlockZ() >= minZ && loc.getBlockZ() <= maxZ) {
                if (!(entity instanceof Player)) {
                    entity.remove();
                }
            }
        }
    }
}

File: src/main/kotlin/me/hhitt/disasters/arena/service/ResetArenaService.kt
package me.hhitt.disasters.arena.service

import com.sk89q.worldedit.WorldEditException
import com.sk89q.worldedit.bukkit.BukkitAdapter
import com.sk89q.worldedit.bukkit.WorldEditPlugin
import com.sk89q.worldedit.extent.clipboard.BlockArrayClipboard
import com.sk89q.worldedit.extent.clipboard.Clipboard
import com.sk89q.worldedit.function.operation.ForwardExtentCopy
import com.sk89q.worldedit.function.operation.Operations
import com.sk89q.worldedit.math.BlockVector3
import com.sk89q.worldedit.regions.CuboidRegion
import com.sk89q.worldedit.session.ClipboardHolder
import me.hhitt.disasters.Disasters
import me.hhitt.disasters.arena.Arena
import me.hhitt.disasters.game.GameState
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.World
import org.bukkit.entity.Player
import java.util.concurrent.CompletableFuture
import kotlin.math.max
import kotlin.math.min

/**
 * ResetArenaService is responsible for resetting the arena to its original state.
 * It saves the current state of the arena and pastes it back when needed.
 *
 * @param arena The arena to be reset.
 * @param worldEdit The WorldEdit plugin instance, if available.
 */

class ResetArenaService(
    private val arena: Arena,
    private val worldEdit: WorldEditPlugin?) {

    private val world = arena.corner1.world
    private lateinit var clipboard: Clipboard
    private lateinit var center: BlockVector3
    private var saveFuture: CompletableFuture<Void>? = null

    private val minX = min(arena.corner1.x, arena.corner2.x).toInt()
    private val maxX = max(arena.corner1.x, arena.corner2.x).toInt()
    private val minY = min(arena.corner1.y, arena.corner2.y).toInt()
    private val maxY = max(arena.corner1.y, arena.corner2.y).toInt()
    private val minZ = min(arena.corner1.z, arena.corner2.z).toInt()
    private val maxZ = max(arena.corner1.z, arena.corner2.z).toInt()


    fun save() {
        val world = arena.corner1.world ?: return println("[Disasters] ERROR: World is null before save!")
        val min: BlockVector3 = BlockVector3.at(arena.corner1.x, arena.corner1.y, arena.corner1.z)
        val max: BlockVector3 = BlockVector3.at(arena.corner2.x, arena.corner2.y, arena.corner2.z)
        val region = CuboidRegion(min, max)
        val blockCount = region.volume
        println("[Disasters] Saving arena '${arena.name}' blocks: $blockCount (async)")

        saveFuture = CompletableFuture.runAsync {
            val clipboard = BlockArrayClipboard(region)
            worldEdit!!.worldEdit.newEditSession(BukkitAdapter.adapt(world)).use { editSession ->
                val forwardExtentCopy = ForwardExtentCopy(editSession, region, clipboard, region.minimumPoint)
                try {
                    Operations.complete(forwardExtentCopy)
                } catch (e: WorldEditException) {
                    throw RuntimeException(e)
                }
                this.clipboard = clipboard
                this.center = region.minimumPoint
            }
            println("[Disasters] Save complete for '${arena.name}'")
        }
    }

    fun waitForSave() {
        saveFuture?.let { future ->
            if (!future.isDone) {
                println("[Disasters] Waiting for save to complete for '${arena.name}'...")
                future.join()
                println("[Disasters] Save complete, starting game for '${arena.name}'")
            }
        }
    }

    fun paste() {
        // Ensure save is complete before pasting
        waitForSave()

        if (!::clipboard.isInitialized) {
            println("[Disasters] ERROR: Clipboard not initialized!")
            return
        }
        println("[Disasters] Paste triggered for '${arena.name}'")
        removeEntitiesInRegion()

        worldEdit!!.worldEdit.newEditSession(BukkitAdapter.adapt(world)).use { editSession ->
            println("[Disasters] Pasting arena '${arena.name}' now...")
            val operation = ClipboardHolder(clipboard)
                .createPaste(editSession)
                .to(center)
                .ignoreAirBlocks(false)
                .build()
            try {
                Operations.complete(operation)
            } catch (e: WorldEditException) {
                throw RuntimeException(e)
            }
        }
        refreshChunks(world, arena.corner1, arena.corner2)
        arena.state = GameState.RECRUITING
        println("[Disasters] Paste complete for '${arena.name}'")
    }

    private fun refreshChunks(world: World, loc1: Location, loc2: Location) {
        val minXX = minX shr 4
        val maxXX = maxX shr 4
        val minZZ = minZ shr 4
        val maxZZ = maxZ shr 4

        for (x in minXX..maxXX) {
            for (z in minZZ..maxZZ) {
                world.refreshChunk(x, z)
            }
        }
    }

    private fun removeEntitiesInRegion() {
        for (entity in world.entities) {
            val loc = entity.location
            if (loc.blockX in minX..maxX && loc.blockY in minY..maxY && loc.blockZ in minZ..maxZ) {
                if (entity !is Player) {
                    entity.remove()
                }
            }
        }
    }


}